### ISSUE TITLE
Define a hyperparameter search space without Raytune 

### ISSUE THREAD

**Issue by L-MASTERS** [ultralytics/ultralytics#8327]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am looking to do hyperparameter tuning on a yolov8 model, and due to the computational resources available to me I don't want to adjust every possible parameter, mainly the initial learning rate, momentum and type of optimizer. I don't want to use Raytune as I have had issues, and this seems common based on other posts here, so I am wondering if there is a way to pass a dictionary of hyperparameters to tune into the tuner class, similar to how you would with Raytune? 

When I tried to do this, I get an error about invalid arguments.

**Comment by Burhan-Q** [comment]
@L-MASTERS the other option for hyperparameter tuning is to use the built-in [mutation algorithm](https://docs.ultralytics.com/guides/hyperparameter-tuning/). As a general caveat (as I don't know your personal level of experience), I like to make sure that others are aware that hyperparameter tuning can take an _exceptionally_ long time to complete, so be sure you're prepared for it to require days, weeks, or even on the scale of months (depending on the search space, number of iterations, number of epochs, dataset size, and hardware) to complete. Additionally, I find that hyperparameter tuning can be helpful to improve performance moderately, but efforts towards cleaning it expanding your dataset are usually going to provide significantly greater improvements to performance.

```python
        from ultralytics import YOLO

        model = YOLO('yolov8n.pt')
        model.tune(space={key1: val1, key2: val2})  # custom search space dictionary
        ```

The format and keys you can use for the `space` argument are:

```python

{  # key: (min, max, gain(optional))

            "lr0": (1e-5, 1e-1),  # initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
            "lrf": (0.0001, 0.1),  # final OneCycleLR learning rate (lr0 * lrf)
            "momentum": (0.7, 0.98, 0.3),  # SGD momentum/Adam beta1
            "weight_decay": (0.0, 0.001),  # optimizer weight decay 5e-4
            "warmup_epochs": (0.0, 5.0),  # warmup epochs (fractions ok)
            "warmup_momentum": (0.0, 0.95),  # warmup initial momentum
            "box": (1.0, 20.0),  # box loss gain
            "cls": (0.2, 4.0),  # cls loss gain (scale with pixels)
            "dfl": (0.4, 6.0),  # dfl loss gain
            "hsv_h": (0.0, 0.1),  # image HSV-Hue augmentation (fraction)
            "hsv_s": (0.0, 0.9),  # image HSV-Saturation augmentation (fraction)
            "hsv_v": (0.0, 0.9),  # image HSV-Value augmentation (fraction)
            "degrees": (0.0, 45.0),  # image rotation (+/- deg)
            "translate": (0.0, 0.9),  # image translation (+/- fraction)
            "scale": (0.0, 0.95),  # image scale (+/- gain)
            "shear": (0.0, 10.0),  # image shear (+/- deg)
            "perspective": (0.0, 0.001),  # image perspective (+/- fraction), range 0-0.001
            "flipud": (0.0, 1.0),  # image flip up-down (probability)
            "fliplr": (0.0, 1.0),  # image flip left-right (probability)
            "mosaic": (0.0, 1.0),  # image mixup (probability)
            "mixup": (0.0, 1.0),  # image mixup (probability)
            "copy_paste": (0.0, 1.0),  # segment copy-paste (probability)
        }
```

**Comment by L-MASTERS** [comment]
@Burhan-Q Thanks for the reply. I have tried to define a parameter space as follows: 
model.tune(data='data.yaml', space={"lr0": (1e-5,1e-1)},epochs=10,iterations=10, batch=8,imgsz=640) 

but I get the following error:
Exception has occurred: SyntaxError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
'[31m[1mspace[0m' is not a valid YOLO argument. Similar arguments are i.e. ['workspace=4', 'save=True', 'patience=50'].

Even if I try just one variable in the space field e.g., 'lr0':1e-5 i get the error.

**Comment by Burhan-Q** [comment]
@L-MASTERS I just tested using Ultralytics `8.1.16` for python 3.10.12 `torch-2.1.0+cu12.1` with this command successfully, so I would make sure to update to a recent version of `pip install -U ultralytics` for compatibility.

```py
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
search = {"lr0":(1e-5, 1e-1), "scale":(0.0, 0.2)}
results = model.tune(
    data="coco8.yaml",
    epochs=3,
    space=search,
    iterations=10,
    imgsz=640,
    batch=4
)
```

You should be able to test with my code as well since I use the `coco8` dataset which is a very small dataset. It took about 5 minutes in total to run on a machine that's relatively old, so it should be quick to check with, even fast if lowering to `iterations=3`.

**Comment by L-MASTERS** [comment]
@Burhan-Q I updated the Ultralytics package and can confirm that it works now. Thanks for the help. Is it possible to vary the optimizer during tuning as well, or should I pick the optimizer that works best from training before starting tuning?

**Comment by Burhan-Q** [comment]
@L-MASTERS glad to hear! I think that the `optimizer` _might be_ tunable with `ray.tune` but not 100% certain. The key-value pair is commented out on this line for the code that uses `ray.tune` https://github.com/ultralytics/ultralytics/blob/2945cfc6ef9e99df2d6589982283d6c68c1d461b/ultralytics/utils/tuner.py#L61

Looks like if uncommented for `ray.tune` it might work (not certain why it's commented tho). Given that it uses `tune.choice()` it seems like this would be _specific_ to `ray.tune` only.

Without `ray.tune` the route like you suggested is probably best, pick an optimizer before kicking off a `tuning` session. The default `"auto"` will try to select an optimizer that's well suited for your training session or pass any of the following 

```py
[
    "SGD",
    "Adam",
    "Adamax",
    "AdamW",
    "NAdam",
    "RAdam",
    "RMSProp",
]
```

to the `optimizer` argument when starting your tuning. FYI, you can also find the list of `optimizer` options in the [`default.yaml` file](https://github.com/ultralytics/ultralytics/blob/2945cfc6ef9e99df2d6589982283d6c68c1d461b/ultralytics/cfg/default.yaml#L24C51-L24C106) if you need a reference. 🚀
